### ISSUE TITLE
error "SWIG wrapped code invalid in 32 bit architecture, regenerate code using -DSWIGWORDSIZE32" 

### ISSUE THREAD

**Issue by haolongzhangm** [swig/swig#1850]
1: on windows with 64 bit , build for x64 ,  so add vcvarsall.bat x64 args, also add CXXFLAGS=-m64 for Ninja

2: so i need enable  -DSWIGWORDSIZE64 for  swig 

then  build will pop error:  error "SWIG wrapped code invalid in 32 bit architecture, regenerate code using -DSWIGWORDSIZE32" 

swig version: https://nchc.dl.sourceforge.net/project/swig/swigwin/swigwin-4.0.2/swigwin-4.0.2.zip 

if I  force add -DSWIGWORDSIZE32 for swig,  build will ok, BUT runtime logic is wrong: exe is bit bit 
file info: _mgb.pyd: PE32+ executable (DLL) (GUI) x86-64, for MS Window,when python  is also a 64bit 
exe, then ptr will be  8 bytes, but when set -DSWIGWORDSIZE32 the wrap.cxx data type is 4 bit , so wrong err

so how to config -DSWIGWORDSIZE64 on windows 64 bit env

**Comment by haolongzhangm** [comment]
any update ?

**Comment by haolongzhangm** [comment]
pls review: https://github.com/swig/swig/pull/1855

**Comment by Mizux** [comment]
you're wrong, on UNIX system std::int64_t are defined as long int, contrary to MSVC thus only here we need to use  `-DSWIGWORDSIZE64` windows take a very conservative approach so you don't need this even in 64bits...

Please take a look at https://github.com/Mizux/cmake-swig/blob/master/doc/swig.md

**Comment by haolongzhangm** [comment]
so uintptr_t size should be 4 or 8 on windows 64?  I think should be 8

BUT at Lib/stdint.i 

 #if defined(SWIGWORDSIZE64)
typedef long int               intptr_t;
typedef unsigned long int      uintptr_t;
 #else
 typedef int                    intptr_t;
 typedef unsigned int           uintptr_t;

both of long int, unsigned long int , int, unsigned int    is 4bit on windows  64

so how to compat this?

**Comment by Mizux** [comment]
good question...
for your information on Archlinux x86_64
```sh
%grepc -rn "typedef.*intptr_t;" /lib/gcc
/lib/gcc/x86_64-pc-linux-gnu/10.1.0/include/stdint-gcc.h:83:typedef __INTPTR_TYPE__ intptr_t;
/lib/gcc/x86_64-pc-linux-gnu/10.1.0/include/stdint-gcc.h:86:typedef __UINTPTR_TYPE__ uintptr_t;
/lib/gcc/x86_64-pc-linux-gnu/10.1.0/include/stdatomic.h:71:typedef _Atomic __INTPTR_TYPE__ atomic_intptr_t;
/lib/gcc/x86_64-pc-linux-gnu/10.1.0/include/stdatomic.h:72:typedef _Atomic __UINTPTR_TYPE__ atomic_uintptr_t;
```
and
```sh
%gcc -dM -E -x c /dev/null | grep __INTPTR
#define __INTPTR_WIDTH__ 64
#define __INTPTR_MAX__ 0x7fffffffffffffffL
#define __INTPTR_TYPE__ long int
```

so using `SWIGWORDSIZE64` is perfect here (as well as for *int64_t typedef) ;)

on MacOS we have (like SWIGWORDSIZE64)
```c
#define __INTPTR_TYPE__ long int
```
but (like SWIGWORDSIZE32)
```c
#define __INT64_TYPE__ long long int
```
which seems close to Windows 64bits

So I suggest to have on MacOS & Windows a new define **SWIGPOINTERSIZE64**
and thus:
```c
/* Types for `void *' pointers.  */
#if defined(SWIGWORDSIZE64) || defined(SWIGPOINTERSIZE64)
typedef long int		intptr_t;
typedef unsigned long int	uintptr_t;
#else
typedef int			intptr_t;
typedef unsigned int		uintptr_t;
#endif
```

so you can enable "64bits" only for pointer type and not for int64_t.

**Comment by Mizux** [comment]
according to `C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\stdint.h`

**Comment by ojwb** [comment]
One point to bear in mind is that SWIG aims to produce code that can then be built on a machine of any architecture (in particular one that's different to the machine SWIG was run on).

Now `stdint.i` is inherently something of a contradiction to that as it needs to care about architecture-specific details - that's the reason for needing the `SWIGWORDSIZE32` and `SWIGWORDSIZE64` defines (rather than just using the type sizes from the machine SWIG is run on).

A problem is that these oversimplify the world into "32 bit" or "64bit" but actually there are certainly more than two different data models - wikipedia lists five (though ILP64 and SILP64 are not common):

https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models

ILP32 is what most modern "32-bit" platforms are (I guess they list it there because it has 64-bit `long long`), but there are also other 32-bit data models such as LP32 (what 16-bit Microsoft Windows used).

It would probably be better to have this specified in terms of codes like `LP64` since those are common terminology - e.g. `SWIGWORDSIZE_LP64` for most modern 64-bit Unix platforms.  That would also provide a natural way to extend to supporting new data models (presumably 128-bit word sizes will appear at some point).
